### PR TITLE
Add GenericZeroMatrix and GenericIdentity class in matrix expressions

### PR DESCRIPTION
**Pull request by asmeurer**
These are generic in the sense that they have no shape (.shape raises a
TypeError). They are returned by MatAdd() and MatMul() with no arguments.

In order to avoid the TypeErrors being an issue, MatAdd and MatMul remove them
unconditionally in their constructors (normally they are completely
unevaluated without calling doit()).

closes #15986 as an alternative. See the discussion there and at
https://github.com/sympy/sympy/pull/15948#issuecomment-463432718.

#### Release Notes

* matrices
  * `MatAdd()` and `MatMul()` now return generic zero and identity matrices with no specified shape.

### CODE REVIEW

**Comment by asmeurer** [comment]
Another option would be to make MatAdd() and MatMul() with no arguments return empty containers as they do now, but add in the aggressive simplification in the constructors as is done here. This would be more or less equivalent to the code in this branch, as it leads to an error if any code happens to run `.shape` on it. I guess the advantage of the code here is that the identity/zero matrices know they are such and so get nice things like printing.

**Comment by asmeurer** [comment]
@mrocklin I'm curious if you have any thoughts on this PR or on https://github.com/sympy/sympy/pull/15986. The goal is to make `MatAdd()` and `MatMul()` with no arguments produce something meaningful, so that code the generates expressions with `f.func(*args)` doesn't fail when `args` is empty.

You wrote in https://github.com/sympy/sympy/issues/2750#issuecomment-32130580 that the matrix expressions should stay "magic free". Do any of these solutions seem magic free to you?

**Comment by mrocklin** [comment]
I'll have to pass on commenting on this one.  My understanding of the
context here is old.

File: sympy/matrices/expressions/tests/test_matexpr.py [hunk 1]
```diff
@@ -357,6 +357,11 @@ def test_issue_2749():
     assert (A.T * A).I.as_explicit() == Matrix([[(A.T * A).I[0, 0], (A.T * A).I[0, 1]], \
     [(A.T * A).I[1, 0], (A.T * A).I[1, 1]]])
 
+
+def test_issue_2750():
```
Comment (jksuom) [inline sympy/matrices/expressions/tests/test_matexpr.py]: [This test](https://github.com/sympy/sympy/blob/master/sympy/matrices/expressions/tests/test_matexpr.py#L360) seems to exist already.

Comment (smichr) [reply sympy/matrices/expressions/tests/test_matexpr.py]: I'm not sure what the comment means. When I click on the link it just takes me to the code which wasn't added in this PR. Are you being thrown off by what sometimes tricks me: the context of the change showing a def that is actually not part of the change?

Comment (jksuom) [reply sympy/matrices/expressions/tests/test_matexpr.py]: Yes, I was wondering why GitHub seems to indicate that it was added in this commit.

**Comment by asmeurer** [comment]
I think the confusion here came from my merge commit, which had this test
as a merge conflict.

**Comment by Upabjojr** [comment]
I think `.shape` should not raise an exception. It should rather return something that matches any number.

**Comment by Upabjojr** [comment]
The scalar values 0 and 1 behave like `GenericZeroMatrix` and `GenericIdentity` mathematically. If you care about the API the latter ones may be useful, though the current API raising exceptions is probably useless.

**Comment by asmeurer** [comment]
That was the approach taken in https://github.com/sympy/sympy/pull/15986 but this approach ended up being cleaner.

**Comment by Upabjojr** [comment]
> That was the approach taken in #15986 but this approach ended up being cleaner.

Well, what about returning `(None, None)` as the shape?

I was reviewing the matrix derivative code and these generic matrices are making the code more complicated. I would prefer `.shape` not to raise exceptions.

**Comment by asmeurer** [comment]
We can reevaluate the best way to do it. Maybe we should combine both approaches. When I made this PR, there weren't any places that were hitting the exception, so it didn't seem like it would be a big deal. But if you are hitting it, then maybe that should be reconsidered.

**Pull request merged**